odxtools/decodestate.py

```diff
@@ -87,32 +93,125 @@ def extract_atomic_value(
-            # bytes to a string
-            text_encoding = "utf-16-be" if is_highlow_byte_order else "utf-16-le"
-            internal_value = internal_value.decode(encoding=text_encoding, errors=text_errors)
+        internal_value: AtomicOdxType
+
+        # Deal with raw byte fields, ...
+        if base_data_type == DataType.A_BYTEFIELD:
+            odxassert(base_type_encoding is None or base_type_encoding == Encoding.NONE)
+
+            internal_value = raw_value
+
+        # ... string types, ...
+        elif base_data_type in (DataType.A_UTF8STRING, DataType.A_ASCIISTRING,
+                                DataType.A_UNICODE2STRING):
+            text_errors = 'strict' if strict_mode else 'replace'
+            if base_type_encoding == Encoding.UTF8:
+                internal_value = raw_value.decode("utf-8", errors=text_errors)
+            elif base_type_encoding == Encoding.UCS2:
+                text_encoding = "utf-16-be" if is_highlow_byte_order else "utf-16-le"
+                internal_value = raw_value.decode(text_encoding, errors=text_errors)
+            elif base_type_encoding == Encoding.ISO_8859_1:
+                internal_value = raw_value.decode("iso-8859-1", errors=text_errors)
+            elif base_type_encoding == Encoding.ISO_8859_2:
+                internal_value = raw_value.decode("iso-8859-2", errors=text_errors)
+            elif base_type_encoding == Encoding.WINDOWS_1252:
+                internal_value = raw_value.decode("cp1252", errors=text_errors)
+            else:
+                odxassert(
+                    base_type_encoding in (None, Encoding.NONE),
+                    f"Specified illegal encoding {base_type_encoding} for string object")
+
+                # if no encoding has been specified explicitly, we
+                # make assumptions by looking at the data type
+                if base_data_type == DataType.A_UTF8STRING:
+                    internal_value = raw_value.decode("utf-8", errors=text_errors)
+                elif base_data_type == DataType.A_UNICODE2STRING:
+                    text_encoding = "utf-16-be" if is_highlow_byte_order else "utf-16-le"
+                    internal_value = raw_value.decode(text_encoding, errors=text_errors)
+                else:
+                    odxassert(base_data_type == DataType.A_ASCIISTRING)
+                    # The spec says ASCII, meaning only character
+                    # values 0-127.  In practice, vendors use
+                    # iso-8859-1, aka latin-1, because iso-8859-1
+                    # never fails since it has a valid character
+                    # mapping for every possible value
+                    internal_value = raw_value.decode("iso-8859-1", errors=text_errors)
+
+        # ... integers, ...
+        elif base_data_type in (DataType.A_INT32, DataType.A_UINT32):
+            if not isinstance(raw_value, int):
+                odxraise(f"Raw value must be of integer type, not {type(raw_value).__name__}")
```

Comment:
Could be a byte array, in case for BCP for example

Reply:
in that case, `bitstruct` would extract a bytes object because we use the format letter `"r"`. since mypy can obviously not understand such semantic behavior, we help it out a bit here...

Reply:
and raw_value would bytes instead of int, no?

Reply:
currently, we always extract an unsigned integer and have to deal with the bit fiddling ourselfs. The reason is that I don't think there is a good pythonic way of converting a bytes string using an exotic encoding (1C, BCD-P, etc) to an integer, even less so if that number is not byte-aligned...

Reply:
That means `if not isinstance(raw_value, int):` will never be true

Reply:
yes, maybe this was a bit too defensive. changed in https://github.com/mercedes-benz/odxtools/pull/381/commits/5c698b549f4dbbb29223b7396bc77d680322c3c1...